contracts/src/gateway/GatewayMessengerFacet.sol

```diff
@@ -20,39 +21,59 @@ contract GatewayMessengerFacet is GatewayActorModifiers {
     using SubnetIDHelper for SubnetID;
 
     /**
-     * @dev sends a general-purpose cross-message from the local subnet to the destination subnet.
+     * @dev Sends a general-purpose cross-message from the local subnet to the destination subnet.
+     * Any value in msg.value will be forwarded in the call.
      *
-     * IMPORTANT: `msg.value` is expected to equal to the value sent in `crossMsg.value` plus the cross-messaging fee.
-     * Only smart contracts are allowed to trigger these cross-net messages, users
-     * can always send funds from their address to the destination subnet and then run the transaction in the destination
-     * normally.
+     * IMPORTANT: Only smart contracts are allowed to trigger these cross-net messages. User wallets can send funds
+     * from their address to the destination subnet and then run the transaction in the destination normally.
      *
-     * @param crossMsg - a cross-message to send.
+     * @param envelope - the original envelope, which will be validated, stamped and committed during the send.
+     * @return committed envelope.
      */
-    function sendContractXnetMessage(IpcEnvelope calldata crossMsg) external payable {
+    function sendContractXnetMessage(
+        IpcEnvelope calldata envelope
+    ) external payable returns (IpcEnvelope memory committed) {
         if (!s.generalPurposeCrossMsg) {
             revert MethodNotAllowed(ERR_GENERAL_CROSS_MSG_DISABLED);
         }
 
-        // we prevent the sender from being an EoA.
+        // TODO: consolidate all these error types into an error InvalidXnetMsg(string reason).
+        //
+        // We prevent the sender from being an EoA.
```

Comment:
To my knowledge, it is not possible to reliably prevent this.

Reply:
In which cases the check below would pass if the sender was a contract? (Note that the sender must exist, so it's either an Eth Account, EVM smart contract, or another built-in actor).

Reply:
That is a very interesting and arguable question in the community. The classic example is code in the constructor.

Reply:
I suspect you're referring to the case where a contract A _under construction_ calls another contract B in its constructor, which then enquires if A is an EOA or a contract based on A's code length. In that case, I think the EVM will report code length = 0 because the EVM bytecode is still being generated. This will give B the impression that A is an EOA. However, I don't see that edge case affecting us here in deleterious ways. We can document the exception that a contract under construction will be prevented from sending xnet messages?